File: backend/api/routes/colg.py
from api.types.colg import HistoryGold, RealTimeGold
from fastapi.params import Param
from api.core.Gzip import register_gzip_request
import json
from fastapi import APIRouter, Path, Query
from api.core.Response import response
from api.core.Redis import get_redis_info
from api.dp import RedisDep
from typing import Annotated
import requests
from config.main import config
from datetime import datetime, timezone, timedelta
from api.core.Response import Return as Response

router = APIRouter()

register_gzip_request(router)

regionMap = {
    '跨1': 1,
    '跨2': 2,
    '跨3A': 3,
    '跨3B': 4,
    '跨4': 5,
    '跨5': 6,
    '跨6': 7,
    '跨7': 8,
}

url = 'https://poservice.colg.cn/api/v1/trend/getCoinTrendDataKol'

payload = {'token': config.COLG_TOKEN, 'area': 0, 'time_type': 'other', 'start_time': '2026-01-06', 'end_time': '2026-01-06'}


def get_date_str(timestamp: int) -> str:
    utc_time = datetime.fromtimestamp(timestamp, tz=timezone.utc)
    beijing_time = utc_time + timedelta(hours=8)
    return beijing_time.strftime('%Y-%m-%d %H:%M:%S')


@router.get('/gold/realtime/', operation_id='realTimeGold')
async def get_real_time_gold(redis: RedisDep) -> Response[list[RealTimeGold]]:
    """
    返回所有跨区的实时金价信息
    """
    key = 'openapi:colg:real_time_gold'
    try:

        def get_gold_data():
            res = []
            for region_name, region_id in regionMap.items():
                payload['area'] = region_id
                payload['start_time'] = datetime.now().strftime('%Y-%m-%d')
                payload['end_time'] = datetime.now().strftime('%Y-%m-%d')
                response = requests.post(url, headers={'Content-Type': 'application/json'}, data=json.dumps(payload))
                response.raise_for_status()
                resp = response.json().get('data', {})
                res.append({'updataTime': get_date_str(resp['x'][-1]), 'region': region_name, 'goldPrice': resp['y'][-1]})
            return res
    except Exception as e:
        # print(f"File not found: {e}")
        return response(data=[])
    # 设置半小时过期一次
    data = get_redis_info(redis, key, get_gold_data, 30 * 60)
    return response(data=data)


@router.get('/gold/history/', operation_id='historyGold')
async def get_history_gold(region: Annotated[int, Query(..., description='跨区id')], timeRange: Annotated[int, Query(..., description='时间跨度天数')] = 7) -> Response[HistoryGold] | Response[None]:
    """
    获取历史金价数据

    Args:
        region (int): 跨区id，必填参数，跨区id对应的映射如下
            {
                '跨1': 1,
                '跨2': 2,
                '跨3A': 3,
                '跨3B': 4,
                '跨4': 5,
                '跨5': 6,
                '跨6': 7,
                '跨7': 8,
            }
        timeRange (int, optional): 时间跨度天数，默认为7天

    Returns:
        Response[HistoryGold]: 返回历史金价数据的响应对象

    Raises:
        无显式异常，但会在跨区ID无效时返回404错误码

    Note:
        - region参数必须在regionMap的值范围内
        - timeRange参数是可选的，使用默认值7表示最近7天的数据
    """
    print(region, timeRange)
    if region not in regionMap.values():
        return response(
            code=404,
            message='跨区ID无效',
            data=None,
        )
    region_name = [k for k, v in regionMap.items() if v == region][0]
    end_date = datetime.now()
    start_date = end_date - timedelta(days=timeRange - 1)
    payload['area'] = region
    payload['start_time'] = start_date.strftime('%Y-%m-%d')
    payload['end_time'] = end_date.strftime('%Y-%m-%d')
    data = []
    echartsData = {}
    try:
        req = requests.post(url, headers={'Content-Type': 'application/json'}, data=json.dumps(payload))
        req.raise_for_status()
        resp = req.json().get('data', {})
        for i in range(len(resp['x'])):
            data.append(
                {
                    'date': get_date_str(resp['x'][i]).split(' ')[0],
                    'avg': resp['y'][i],
                    'zero': resp['y1'][i],
                }
            )
    except Exception as e:
        return response(data=None)
    echartsData = {
        'title': {'text': f'DNF{region_name}近{timeRange}天金币价格趋势分析', 'left': 'center'},
        'toolbox': {'feature': {'saveAsImage': {}, 'dataView': {'readOnly': True}, 'magicType': {'type': ['line', 'bar']}}},
        'tooltip': {'trigger': 'axis', 'axisPointer': {'type': 'cross', 'label': {'backgroundColor': '#6a7985'}}},
        'legend': {'data': ['平均金价', '12时金价'], 'bottom': 0},
        'xAxis': {'type': 'category', 'name': '日期', 'boundaryGap': False, 'data': [item['date'] for item in data]},
        'yAxis': {'type': 'value', 'name': '价格 (万金币/1人民币)', 'min': (min([item['avg'] for item in data] + [item['zero'] for item in data]) // 5 ) * 5, 'max': (max([item['avg'] for item in data] + [item['zero'] for item in data]) // 5 ) * 5 + 5, 'axisLabel': {'formatter': '{value} 万金币'}},
        'series': [
            {'name': '平均金价', 'type': 'line', 'smooth': True, 'data': [item['avg'] for item in data], 'markPoint': {'data': [{'type': 'max', 'name': '最大值'}, {'type': 'min', 'name': '最小值'}]}},
            {'name': '12时金价', 'type': 'line', 'smooth': True, 'data': [item['zero'] for item in data], 'markLine': {'data': [{'type': 'average', 'name': '平均值'}]}},
        ],
    }
    return response(data={'region': region_name, 'data': data, 'echartsData': echartsData})

File: backend/go/internal/service/colg.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"dcalc-go/internal/config"
)

type ColgService struct {
	cache  *Cache
	config config.Config
	client *http.Client
}

var regionMap = map[string]int{
	"跨1":  1,
	"跨2":  2,
	"跨3A": 3,
	"跨3B": 4,
	"跨4":  5,
	"跨5":  6,
	"跨6":  7,
	"跨7":  8,
}

func NewColgService(cache *Cache, cfg config.Config) *ColgService {
	return &ColgService{
		cache:  cache,
		config: cfg,
		client: &http.Client{Timeout: 12 * time.Second},
	}
}

func (s *ColgService) RealTimeGold(ctx context.Context) ([]map[string]any, error) {
	if s.config.ColgToken == "" {
		return nil, fmt.Errorf("COLG_TOKEN is required")
	}

	return GetOrLoad(s.cache, "colg:realtime", 30*time.Minute, func() ([]map[string]any, error) {
		result := make([]map[string]any, 0, len(regionMap))
		now := time.Now().Format("2006-01-02")
		for regionName, regionID := range regionMap {
			data, err := s.requestTrendData(ctx, regionID, now, now)
			if err != nil {
				return nil, err
			}
			xValues, _ := data["x"].([]any)
			yValues, _ := data["y"].([]any)
			if len(xValues) == 0 || len(yValues) == 0 {
				continue
			}
			result = append(result, map[string]any{
				"updataTime": formatBeijingTime(int64FromAny(xValues[len(xValues)-1])),
				"region":     regionName,
				"goldPrice":  float64FromAny(yValues[len(yValues)-1]),
			})
		}
		return result, nil
	})
}

func (s *ColgService) HistoryGold(ctx context.Context, region int, timeRange int) (map[string]any, error) {
	if s.config.ColgToken == "" {
		return nil, fmt.Errorf("COLG_TOKEN is required")
	}
	if timeRange <= 0 {
		timeRange = 7
	}

	regionName := ""
	for name, id := range regionMap {
		if id == region {
			regionName = name
			break
		}
	}
	if regionName == "" {
		return nil, fmt.Errorf("invalid region")
	}

	endDate := time.Now()
	startDate := endDate.AddDate(0, 0, -(timeRange - 1))
	trend, err := s.requestTrendData(ctx, region, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))
	if err != nil {
		return nil, err
	}

	xValues, _ := trend["x"].([]any)
	yValues, _ := trend["y"].([]any)
	y1Values, _ := trend["y1"].([]any)
	rows := make([]map[string]any, 0, len(xValues))
	minValue := 0.0
	maxValue := 0.0
	for index := range xValues {
		avg := float64FromAny(yValues[index])
		zero := float64FromAny(y1Values[index])
		if index == 0 || avg < minValue {
			minValue = avg
		}
		if zero < minValue {
			minValue = zero
		}
		if index == 0 || avg > maxValue {
			maxValue = avg
		}
		if zero > maxValue {
			maxValue = zero
		}
		rows = append(rows, map[string]any{
			"date": strings.Split(formatBeijingTime(int64FromAny(xValues[index])), " ")[0],
			"avg":  avg,
			"zero": zero,
		})
	}

	title := fmt.Sprintf("DNF%s近%d天金币价格趋势分析", regionName, timeRange)
	echarts := map[string]any{
		"title":   map[string]any{"text": title, "left": "center"},
		"toolbox": map[string]any{"feature": map[string]any{"saveAsImage": map[string]any{}, "dataView": map[string]any{"readOnly": true}, "magicType": map[string]any{"type": []string{"line", "bar"}}}},
		"tooltip": map[string]any{"trigger": "axis", "axisPointer": map[string]any{"type": "cross", "label": map[string]any{"backgroundColor": "#6a7985"}}},
		"legend":  map[string]any{"data": []string{"平均金价", "12时金价"}, "bottom": 0},
		"xAxis":   map[string]any{"type": "category", "name": "日期", "boundaryGap": false, "data": extractField(rows, "date")},
		"yAxis":   map[string]any{"type": "value", "name": "价格 (万金币/1人民币)", "min": floorToFive(minValue), "max": floorToFive(maxValue) + 5, "axisLabel": map[string]any{"formatter": "{value} 万金币"}},
		"series": []map[string]any{
			{"name": "平均金价", "type": "line", "smooth": true, "data": extractField(rows, "avg"), "markPoint": map[string]any{"data": []map[string]any{{"type": "max", "name": "最大值"}, {"type": "min", "name": "最小值"}}}},
			{"name": "12时金价", "type": "line", "smooth": true, "data": extractField(rows, "zero"), "markLine": map[string]any{"data": []map[string]any{{"type": "average", "name": "平均值"}}}},
		},
	}

	return map[string]any{"region": regionName, "data": rows, "echartsData": echarts}, nil
}

func (s *ColgService) OfficialGold(ctx context.Context) ([]map[string]any, error) {
	if s.config.HelperToken == "" || s.config.HelperID == "" {
		return nil, fmt.Errorf("HELPER_TOKEN and HELPER_ID are required")
	}

	return GetOrLoad(s.cache, "colg:official", 30*time.Minute, func() ([]map[string]any, error) {
		servers := []string{"1", "2", "3A", "3B", "4", "5", "6", "7"}
		result := make([]map[string]any, 0, len(servers))
		for _, server := range servers {
			data, err := s.requestOfficialGold(ctx, server)
			if err != nil {
				return nil, err
			}
			result = append(result, data)
		}
		return result, nil
	})
}

func (s *ColgService) requestTrendData(ctx context.Context, region int, startTime string, endTime string) (map[string]any, error) {
	payload := map[string]any{
		"token":      s.config.ColgToken,
		"area":       region,
		"time_type":  "other",
		"start_time": startTime,
		"end_time":   endTime,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://poservice.colg.cn/api/v1/trend/getCoinTrendDataKol", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var payloadResp map[string]any
	if err := json.Unmarshal(content, &payloadResp); err != nil {
		return nil, err
	}
	data, _ := payloadResp["data"].(map[string]any)
	return data, nil
}

func (s *ColgService) requestOfficialGold(ctx context.Context, server string) (map[string]any, error) {
	form := url.Values{}
	form.Set("r", "godRank/godDealRank")
	form.Set("userId", s.config.HelperID)
	form.Set("token", s.config.HelperToken)
	form.Set("serverNo", server)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://dzhu.qq.com/zangyi/activity/app", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var payloadResp map[string]any
	if err := json.Unmarshal(content, &payloadResp); err != nil {
		return nil, err
	}
	data, _ := payloadResp["data"].(map[string]any)
	return map[string]any{
		"updataTime":      time.Now().Format("2006-01-02 15:04:05"),
		"region":          "跨" + server,
		"goldDealRank":    extractPrices(data["godDealRank"]),
		"goldTopDealRank": extractPrices(data["godTopDealRank"]),
	}, nil
}

func extractPrices(value any) []int {
	rows, _ := value.([]any)
	result := make([]int, 0, len(rows))
	for _, row := range rows {
		rowMap, _ := row.(map[string]any)
		result = append(result, intFromAnyString(rowMap["price"]))
	}
	return result
}

func extractField(rows []map[string]any, key string) []any {
	result := make([]any, 0, len(rows))
	for _, row := range rows {
		result = append(result, row[key])
	}
	return result
}

func formatBeijingTime(ts int64) string {
	return time.Unix(ts, 0).UTC().Add(8 * time.Hour).Format("2006-01-02 15:04:05")
}

func floorToFive(value float64) float64 {
	return float64(int(value/5) * 5)
}

func float64FromAny(value any) float64 {
	switch typed := value.(type) {
	case float64:
		return typed
	case float32:
		return float64(typed)
	case int:
		return float64(typed)
	case int64:
		return float64(typed)
	case json.Number:
		parsed, _ := typed.Float64()
		return parsed
	default:
		return 0
	}
}

func int64FromAny(value any) int64 {
	switch typed := value.(type) {
	case int64:
		return typed
	case int:
		return int64(typed)
	case float64:
		return int64(typed)
	case json.Number:
		parsed, _ := typed.Int64()
		return parsed
	default:
		return 0
	}
}

func intFromAnyString(value any) int {
	switch typed := value.(type) {
	case string:
		parsed, _ := strconv.Atoi(typed)
		return parsed
	default:
		return int(float64FromAny(value))
	}
}
